### ISSUE TITLE
Checking for duplicate records

### ISSUE THREAD

**Issue by jeremyshelton** [kartoza/django-bims#1797]
@dimasciput @timlinux 
Following on from our chat yesterday, it would be useful to create a tool that allows us to check for duplicate records. For example, if a fish record has the same location, date and taxon, then it is in all likelihood a duplicate. 
@helendallas

**Comment by helendallas** [comment]
@jeremyshelton Biotope would also need to be included.

I agree this would be a good tool, but it needs to include all levels leading to the occurrence record to ensure only true duplicates are reported

**Comment by dimasciput** [comment]
@Fanevanjanahary please chat with me first before working on this one.

**Comment by timlinux** [comment]
Also we should use biotope for uniqueness check so that we take into account the fact that the same organism can be found in the same site and date but in different habitats.

**Comment by helendallas** [comment]
@Fanevanjanahary 

So the hierarchy for checking for duplicates is:

Site Code - Sampling Date - Biotope - Taxon

Biotope is the aquatic habitat that the fish or invertebrate is sampled in. They are in BIMS- Biotopes. The BIotope includes three levels and I think it should be a combination of **Broad - Specific - Substratum** to ensure only real duplicates are included.

**Comment by jeremyshelton** [comment]
@Fanevanjanahary @dimasciput would you like us to test this now, or are you still working on it?

**Comment by dimasciput** [comment]
@jeremyshelton you can test it in the testing site

**Comment by jeremyshelton** [comment]
@dimasciput @Fanevanjanahary thanks - where is this tool?
...don't see it in the Administration tab?

**Comment by Fanevanjanahary** [comment]
@jeremyshelton you can find it in summary report
 https://testing.healthyrivers.kartoza.com/summary-report/

**Comment by jeremyshelton** [comment]
Thanks @Fanevanjanahary 
For future reference: how and where can we generate a summary report?

**Comment by helendallas** [comment]
@jeremyshelton It is is adminstration/summary report.

The worrying thing is there are 

![image](https://user-images.githubusercontent.com/45286876/126746581-08a91be0-57ca-4701-8ab3-6f167ae14c3c.png)


Also there is no way to fix these.....we need further guidance and discussion on this @dimasciput @Fanevanjanahary

**Comment by jeremyshelton** [comment]
OK thanks @helendallas 
Trying to download the .csv file, but it is stuck on 'processing'?

**Comment by helendallas** [comment]
@dimasciput @Fanevanjanahary 
Please can we discuss this at our meeting tomorrow. This tool needs to be accessible from the Admin dropdown, and we need to be able to check and then fix duplicate records, so the workflow needs to :
1. Identify potential duplicate records
2. Allow us to be able to fix these records
Perhaps I do not yet understand the workflow, so please can you talk us through this Dimas.

**Comment by dimasciput** [comment]
@Fanevanjanahary please download the duplicate records, and share it here.

**Comment by Fanevanjanahary** [comment]
I Will share, It takes a long time

**Comment by Fanevanjanahary** [comment]
Can't upload CSV here so I put it on NextCloud DropZone/FBIS

**Comment by helendallas** [comment]
@Fanevanjanahary How do I access it? Can you share with Wetransfer?

https://wetransfer.com/

**Comment by Fanevanjanahary** [comment]
@helendallas please find here https://we.tl/t-rdAH4XNmTl

**Comment by helendallas** [comment]
@Fanevanjanahary @dimasciput I cant figure out which are the duplicate records. To do this we need a csv file that contains the following:

- FBIS Site Code
- Site description
- Latitude
- Longitude
- Sampling Date
- Taxon
- Broad Biotope
- Specific Biotope
- Substratum
- Sampling method
- Abundance measure
- Abundance value
- Collector/Owner
- Collector/Owner Institute
- Author(s)
- Year
- Source
- Title
- Notes

At the moment Site is a number.

@dimasciput We need the data returned as close to the Occurrence upload file as possible and the somewhere there needs to be a flag that it is a duplicate record.

Also some Institution IDs, healthyrivers????? and other strange columns and numbers which are difficult for us to follow

@Fanevanjanahary Please can you also clarify what you included in the criteria to check for duplicates. I doubt that all 70,000 records are duplicates.

**Comment by Fanevanjanahary** [comment]
@helendallas if they are in the same  `site`   also if the value of `date`,  `biotope`,  `specific_biotope`,  `substratum`, `taxonomy`,
            `abundance_number` are same

**Comment by helendallas** [comment]
@Fanevanjanahary Please can you share the duplicates files again. Sorry I didn't download it in time

**Comment by Fanevanjanahary** [comment]
@helendallas you can download it here  https://we.tl/t-WBogEY0Zlh

**Comment by helendallas** [comment]
As per our sprint planning meeting @dimasciput to check.

**Comment by jeremyshelton** [comment]
@timlinux @dimasciput @Fanevanjanahary Can we please pick this up as a priority at the next sprint.
Thanks
@helendallas

**Comment by jeremyshelton** [comment]
@Fanevanjanahary @dimasciput I cant figure out which are the duplicate records. To do this we need a csv file that contains the following:

- FBIS Site Code
- Site description
- Latitude
- Longitude
- Sampling Date
- Taxon
- Broad Biotope
- Specific Biotope
- Substratum
- Sampling method
- Abundance measure
- Abundance value
- Collector/Owner
- Collector/Owner Institute
- Author(s)
- Year
- Source
- Title
- Notes

At the moment Site is a number.

@dimasciput We need the data returned as close to the Occurrence upload file as possible and the somewhere there needs to be a flag that it is a duplicate record.

Also some Institution IDs, healthyrivers????? and other strange columns and numbers which are difficult for us to follow

**Comment by dimasciput** [comment]
Hi @jeremyshelton @helendallas , I will try to clean up some duplicate I can find using a script first, and then if there are still some duplicates left I will give you a csv file to check and investigate. Hopefully it will be ready by tomorrow or Thursday.

**Comment by timlinux** [comment]
Yes I think we will probably need to allocate around 3 days to deal with this programmatically. I am thinking we could perhaps take a different approach and include metabase in our stack which will let us make ad hoc, free form queries. It is very user friendly and does not need a programmer to be able to gain insights like listing duplicates etc.

**Comment by jeremyshelton** [comment]
Hi @timlinux @dimasciput 

This is a feature request that has come up repeatedly from the users, and so we would really like to prioritise this ticket. 
Lets discuss exactly what's needed to get this done at our next meet-up.
@helendallas

**Comment by helendallas** [comment]
@dimasciput Please five us an update on the progress on this ticket

**Comment by helendallas** [comment]
@dimasciput any more progress on this ticket?
